### PR TITLE
Add overridable attributes to all types (defaulting to empty) 

### PR DESCRIPTION
**Pull request by Kangz**
@litherum PTAL at the second commit, would this work for you?

### CODE REVIEW

**Comment by austinEng** [comment]
while you're at it; you may consider adding a WGPU_NULLABLE attribute too?

**Comment by Kangz** [comment]
I'll add `WGPU_NULLABLE` if we decide to move forward after the discussions in #179

**Comment by litherum** [comment]
+1 to nullable <3

Yes, this patch looks perfect.

**Review by litherum — APPROVED** [review]

**Comment by Kangz** [comment]
PTAL again, I had to bundle a couple changes that would have been annoying to split up.

File: webgpu.h [hunk 1]
```diff
@@ -723,29 +818,29 @@ typedef struct WGPUCompilationMessage {
...
 typedef struct WGPUComputePassTimestampWrite {
     WGPUQuerySet querySet;
     uint32_t queryIndex;
     WGPUComputePassTimestampLocation location;
-} WGPUComputePassTimestampWrite;
+} WGPUComputePassTimestampWrite WGPU_STRUCTURE_ATTRIBUTE;
 
 typedef struct WGPUConstantEntry {
     WGPUChainedStruct const * nextInChain;
     char const * key;
```
Comment (kainino0x) [inline webgpu.h:831]: we might also want a NONNULL for all of the pointers that aren't nullable?

File: webgpu.h [hunk 2]
```diff
@@ -48,6 +48,22 @@
...
+#define WGPU_ENUM_ATTRIBUTE
+#endif
+#if !defined(WGPU_STRUCTURE_ATTRIBUTE)
+#define WGPU_STRUCTURE_ATTRIBUTE
+#endif
+#if !defined(WGPU_FUNCTION_ATTRIBUTE)
+#define WGPU_FUNCTION_ATTRIBUTE
+#endif
+#if !defined(WGPU_NULLABLE)
+#define WGPU_NULLABLE
```
Comment (kainino0x) [inline webgpu.h:64]: Do we want the header to use the clang extension (and maybe other extensions) by default if available, or just leave this for people who want to take advantage of it?

File: webgpu.h [hunk 3]
```diff
@@ -639,18 +723,29 @@ typedef enum WGPUTextureUsage {
...
+typedef void (*WGPUCreateRenderPipelineAsyncCallback)(WGPUCreatePipelineAsyncStatus status, WGPURenderPipeline pipeline, char const * message, void * userdata) WGPU_FUNCTION_ATTRIBUTE;
+typedef void (*WGPUDeviceLostCallback)(WGPUDeviceLostReason reason, char const * message, void * userdata) WGPU_FUNCTION_ATTRIBUTE;
+typedef void (*WGPUErrorCallback)(WGPUErrorType type, char const * message, void * userdata) WGPU_FUNCTION_ATTRIBUTE;
+typedef void (*WGPUProc)(void) WGPU_FUNCTION_ATTRIBUTE;
+typedef void (*WGPUQueueWorkDoneCallback)(WGPUQueueWorkDoneStatus status, void * userdata) WGPU_FUNCTION_ATTRIBUTE;
+typedef void (*WGPURequestAdapterCallback)(WGPURequestAdapterStatus status, WGPUAdapter adapter, char const * message, void * userdata) WGPU_FUNCTION_ATTRIBUTE;
+typedef void (*WGPURequestDeviceCallback)(WGPURequestDeviceStatus status, WGPUDevice device, char const * message, void * userdata) WGPU_FUNCTION_ATTRIBUTE;
 
 typedef struct WGPUChainedStruct {
     struct WGPUChainedStruct const * next;
```
Comment (kainino0x) [inline webgpu.h:741]: These should be nullable

**Review by kainino0x — APPROVED** [review]
LGTM w/ non-blocking comments

File: webgpu.h [hunk 4]
```diff
@@ -149,11 +149,14 @@ typedef enum WGPUBufferBindingType {
...
     WGPUBufferMapAsyncStatus_Success = 0x00000000,
-    WGPUBufferMapAsyncStatus_Error = 0x00000001,
+    WGPUBufferMapAsyncStatus_ValidationError = 0x00000001,
     WGPUBufferMapAsyncStatus_Unknown = 0x00000002,
     WGPUBufferMapAsyncStatus_DeviceLost = 0x00000003,
     WGPUBufferMapAsyncStatus_DestroyedBeforeCallback = 0x00000004,
     WGPUBufferMapAsyncStatus_UnmappedBeforeCallback = 0x00000005,
+    WGPUBufferMapAsyncStatus_MappingAlreadyPending = 0x00000006,
+    WGPUBufferMapAsyncStatus_OffsetOutOfRange = 0x00000007,
+    WGPUBufferMapAsyncStatus_SizeOutOfRange = 0x00000008,
```
Comment (kainino0x) [inline webgpu.h:247]: We did this for Dawn but Austin(?) has pointed out that the mapAsync callback should maybe just have a `const char * message` instead, so not sure we should land it here yet

Comment (Kangz) [reply webgpu.h:247]: This has already landed in https://github.com/webgpu-native/webgpu-headers/pull/181 but we could discuss further and change it!

File: webgpu.h (hunk 1 above)
Comment (Kangz) [reply webgpu.h:831]: That's would be a lot of noise, but maybe? What do others think of adding a `WGPU_NONNULL`

Comment (kainino0x) [reply webgpu.h:831]: @litherum ?

File: webgpu.h (hunk 4 above)
Comment (kainino0x) [reply webgpu.h:247]: SG, we'll worry about it later.

**Comment by kainino0x** [comment]
Rebased past #181, let's do the rest as followup

**Pull request merged**